packages/server/dataloader/primaryKeyLoaderMaker.ts

```diff
@@ -20,3 +20,20 @@ export function primaryKeyLoaderMaker<ReturnT extends {id: string}>(
     )
   }
 }
+
+/**
+ * Register a loader for types loaded by their {id: number} field.
+ */
+export function numberedPrimaryKeyLoaderMaker<ReturnT extends {id: number}>(
```

Comment:
I had hard time merging this with the one above...

Reply:
We should make the key type generic in `primaryKeyLoaderMaker` instead:
```ts
export function primaryKeyLoaderMaker<KeyT extends string | number, ReturnT extends {id: KeyT}>(
  batchFn: (ids: readonly KeyT[]) => Promise<ReturnT[]>
) {
  return (parent: RootDataLoader) => {
    return new NullableDataLoader<KeyT, ReturnT, KeyT>(
      async (ids) => {
        const result = await batchFn(ids)
        return normalizeResults(ids, result)
      },
      {...parent.dataLoaderOptions}
    )
  }
}
```
I also prepared already the `normalizeResults` function but didn't had any use yet, now it's there:
```ts
export const normalizeResults = <KeyT extends string | number, T extends {[key: string]: KeyT}>(
  keys: Readonly<KeyT[]>,
  results: T[],
  key: keyof T = 'id'
) => {
  const map = {} as Record<KeyT, T>
  results.forEach((result: T) => {
    result[key]
    map[result[key]] = result
  })
  const mappedResults = [] as T[]
  keys.forEach((key) => {
    mappedResults.push(map[key])
  })
  return mappedResults
}
```

I don't know where exactly the type gets lost in `foreignKeyLoaderMaker` but sprinkling in some more helper types and strategic casting got it working
```ts
type LoaderKeyType<LoaderName extends LoaderKeys> = Loader<LoaderName> extends DataLoader<
  infer KeyT,
  any,
  infer KeyT
>
  ? KeyT : any

/**
 * Used to register loaders for types by foreign key.
 *
 * This is intended for types which already have a {@link primaryKeyLoaderMaker} defined.
 * When an item is loaded via this loader, the primary loader will be primed with the result as well.
 * It reflects a one to many relationship, i.e. for each key passed, an array will be returned.
 */
export function foreignKeyLoaderMaker<
  LoaderName extends LoaderKeys,
  KeyName extends keyof LoaderType<LoaderName>
>(
  primaryLoaderKey: LoaderName,
  foreignKey: KeyName,
  fetchFn: (keys: readonly LoaderType<LoaderName>[KeyName][]) => Promise<LoaderType<LoaderName>[]>
) {
  type T = LoaderType<LoaderName>
  type PrimaryKeyT = LoaderKeyType<LoaderName>
  type KeyValue = T[KeyName]
  return (parent: RootDataLoader) => {
    const primaryLoader = parent.get(primaryLoaderKey) as DataLoader<PrimaryKeyT, T, PrimaryKeyT>
    return new DataLoader<KeyValue, T[], KeyValue>(
      async (ids) => {
        const items = await fetchFn(ids)
        items.forEach((item) => {
          const key = item?.['id']
          if (key) {
            primaryLoader.clear(key).prime(key, item)
          }
        })
        return ids.map((id) => items.filter((item) => item && item[foreignKey] === id))
      },
      {
        ...parent.dataLoaderOptions
      }
    )
  }
}
```
Basically the `parent.get(primaryLoaderKey)` seems to be the culprit. Some type is not restrictive enough to make it work. Not very satisfactory but it should work.